borsh/src/de/mod.rs

```diff
@@ -144,6 +145,92 @@ impl BorshDeserialize for usize {
     }
 }
 
+mod non_zero {
+    use super::*;
+    use core::num::*;
+
+    trait NonZero {
+        type Primitive;
+    }
+
+    macro_rules! non_zero_primitives {
+        ($($type:ty => $primitive:ty,)+) => {
+            $(
+                impl NonZero for $type {
+                    type Primitive = $primitive;
+                }
+            )+
+        }
+    }
+
+    non_zero_primitives! {
+        NonZeroU8   => u8,
+        NonZeroU16  => u16,
+        NonZeroU32  => u32,
+        NonZeroU64  => u64,
+        NonZeroU128 => u128,
+
+        NonZeroI8   => i8,
+        NonZeroI16  => i16,
+        NonZeroI32  => i32,
+        NonZeroI64  => i64,
+        NonZeroI128 => i128,
+
+        NonZeroUsize => usize,
+    }
+
+    macro_rules! impl_for_non_zero_integer {
+        ($type:ty) => {
+            impl BorshDeserialize for $type {
+                #[inline]
+                fn deserialize(buf: &mut &[u8]) -> Result<Self> {
+                    let value: <$type as NonZero>::Primitive = <_>::deserialize(buf)?;
+                    <$type>::new(value)
+                        .ok_or_else(|| Error::new(ErrorKind::InvalidInput, ERROR_NON_ZERO_EXPECTED))
```

Comment:
I'm not sure `InvalidInput` or `InvalidData` error kind should be used there.

Reply:
I feel `InvalidInput` is fine, we use it in `bool` deserialization as well, but thanks for raising the concern :+1:

Reply:
Good point: we should be consistent first and foremost,  but indeed it'd be cool to fix separaetely (definitelly not in this PR : https://github.com/near/borsh-rs/issues/48)